pkg/controller/azvolumeattachment.go

```diff
@@ -147,64 +149,102 @@ func (r *reconcileAzVolumeAttachment) GetReplicaCount(ctx context.Context, under
 }
 
 // ManageAttachmentsForVolume will be runing on a separate channel
-func (r *reconcileAzVolumeAttachment) SyncAll(ctx context.Context) error {
+func (r *reconcileAzVolumeAttachment) SyncAll(ctx context.Context) (bool, error) {
 	r.syncMutex.Lock()
 	defer r.syncMutex.Unlock()
 
-	// Get all AzVolume
-	azVolumes, err := r.azVolumeClient.DiskV1alpha1().AzVolumes(r.namespace).List(ctx, metav1.ListOptions{})
-	if err != nil {
-		klog.Errorf("failed to list AzVolumes: %v", err)
-		return err
+	volumesToSync := []string{}
+	// Get all volumeAttachments
+	var volumeAttachments storagev1.VolumeAttachmentList
+	if err := r.client.List(ctx, &volumeAttachments, &client.ListOptions{}); err != nil {
+		return true, err
+	}
+
+	// Loop through volumeAttachments and create Primary AzVolumeAttachments in correspondence
+	for _, volumeAttachment := range volumeAttachments.Items {
+		if volumeAttachment.Spec.Attacher == azureutils.DriverName {
+			volumeName := volumeAttachment.Spec.Source.PersistentVolumeName
+			if volumeName == nil {
+				continue
+			}
+			volumesToSync = append(volumesToSync, *volumeName)
+
+			nodeName := volumeAttachment.Spec.NodeName
+			err := r.client.Create(ctx, &v1alpha1.AzVolumeAttachment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: azureutils.GetAzVolumeAttachmentName(*volumeName, nodeName),
+				},
+				Spec: v1alpha1.AzVolumeAttachmentSpec{
+					UnderlyingVolume: *volumeName,
+					NodeName:         nodeName,
+					RequestedRole:    v1alpha1.PrimaryRole,
+				},
+			}, &client.CreateOptions{})
+			if err != nil {
```

Comment:
If there are a number of VolumeAttachment objects and the driver fails to create one of them, this loop will return true. This will trigger a retry and we will loop through the entire list of VolumeAttachments all over again. Shouldn't we only try to create the AzVolumeAttachment objects for only the failed ones while retrying?

Reply:
Good point 👍